### PR TITLE
Add retry proxy wrapper for ex-db defaults

### PR DESCRIPTION
**Pull request by pivnicek**
fixes: #125

### CODE REVIEW

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        ?int $maxTries = null,
+        ?array $expectedExceptions = null,
+        ?int $backoffInterval = null,
+        ?RetryPolicyInterface $retryPolicy = null,
+        ?BackoffPolicyInterface $backoffPolicy = null
+    ) {
+        if ($retryPolicy === null) {
+            $retryPolicy = new SimpleRetryPolicy(
+                $maxTries ?? self::DEFAULT_MAX_TRIES,
+                $expectedExceptions ? $expectedExceptions : self::DEFAULT_EXCEPTED_EXCEPTIONS
```
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: Tady by taky mohly být `??`

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            );
+        }
+        if ($backoffPolicy === null) {
+            $backoffPolicy = new ExponentialBackOffPolicy(
+                $backoffInterval ? $backoffInterval : self::DEFAULT_BACKOFF_INTERVAL
```
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: Tady by taky mohly být `??`

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 1 above)
Comment (tomasfejfar) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: Akorát to povolí []

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 2 above)
Comment (tomasfejfar) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: Akorát to povolí 0

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        LoggerInterface $logger,
+        ?int $maxTries = null,
+        ?array $expectedExceptions = null,
+        ?int $backoffInterval = null,
+        ?RetryPolicyInterface $retryPolicy = null,
+        ?BackoffPolicyInterface $backoffPolicy = null
+    ) {
+        if ($retryPolicy === null) {
+            $retryPolicy = new SimpleRetryPolicy(
+                $maxTries ?? self::DEFAULT_MAX_TRIES,
```
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: Tohle povoluje 0, což je asi správně.

Comment (tomasfejfar) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: Nejsem si jistej, jak to vlastně je. Jestli to udělá jednou a tedy "max tries" je vlastně spíš "max retries". Udělá to a pak 0 až XX krát to zkusí retrynout. Nebo dostane ukol a zkusí ho 1 až XX krát udělat. 

Tj. když je "maxTries" == 0, spustí se to vůbec někdy?

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+namespace Keboola\DbExtractor;
+
+use Psr\Log\LoggerInterface;
+use Retry\RetryProxy;
+use Retry\Policy\RetryPolicyInterface;
+use Retry\Policy\SimpleRetryPolicy;
+use Retry\BackOff\BackOffPolicyInterface;
+use Retry\BackOff\ExponentialBackOffPolicy;
+
+class DbRetryProxy extends RetryProxy
```
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: Není důvod, aby k tomu nebyly testy.

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 1 above)
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: ano, mas pravdu

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 2 above)
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: jop, dik

**Comment by ondrajodas** [comment]
jak to vypadá? mohli bychom to dotáhnout abychom mohli pokračovat dále?

**Comment by pivnicek** [comment]
Promin Ondrej za pozdenni

**Comment by pivnicek** [comment]
@tomasfejfar can you please take a look?  I actually think the class should be renamed to just RetryProxy so that the current extracor codes wont need updating

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 4 above)
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: I don't think it should be named the same, just from different namespace. That may cause some confusion. You'll need to change the "use" statements anyway, so it's not much help IMHO. While making it "Db" prefixed sends a clear message that this is no ordinary RetryProxy.

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+{
+    public const DEFAULT_MAX_TRIES = 5;
+
+    /* set initial backoff to 1 second */
+    public const DEFAULT_BACKOFF_INTERVAL = 1000;
+
+    public const DEFAULT_EXCEPTED_EXCEPTIONS = ['PDOException', 'ErrorException'];
+
+    public function __construct(
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: sure, it's BC breaking to rename it anyways, so we can change the order.

File: tests/DbRetryProxyTest.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        $res = $retryProxy->call(function () use (&$i): int {
+            $i++;
+            if ($i < 5) {
+                throw new \PDOException('test throw ' . $i);
+            } else {
+                return $i;
+            }
+        });
+        $this->assertEquals(5, $res);
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply tests/DbRetryProxyTest.php]: I suppose,  I had intended to, that's why I set it up :-).

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 4 above)
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: ok, we'll have to update the extractors, but we're doing that anyways.  cool, agreed.

File: src/Keboola/DbExtractor/DbRetryProxy.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: hmmm,  I don't think I agree.  here for example https://github.com/keboola/db-extractor-common/pull/126/files/3fe8ef6155f9da10355b1ef1213b60318a8bd523#diff-262b537e49b5aac9a5d12edf71cf7b13R200

[the remark this answers is not in the document]
Comment (pivnicek) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: plus all the usages are already in that order (in the extractors)

**Comment by pivnicek** [comment]
ok @tomasfejfar I updated to your suggestions except for the argument order change.  I'd like to keep it as is.

File: src/Keboola/DbExtractor/DbRetryProxy.php [hunk 7]
```diff
@@ -18,7 +18,7 @@ class DbRetryProxy extends RetryProxy
     /* set initial backoff to 1 second */
     public const DEFAULT_BACKOFF_INTERVAL = 1000;
 
-    public const DEFAULT_EXCEPTED_EXCEPTIONS = ['PDOException', 'ErrorException'];
+    public const DEFAULT_EXPECTED_EXCEPTIONS = ['PDOException', 'ErrorException'];
```
Comment (tomasfejfar) [inline src/Keboola/DbExtractor/DbRetryProxy.php]: Haha, I thought that's weird, but thought that it's like "throw on everything EXCEPT those" :D

Comment (tomasfejfar) [reply src/Keboola/DbExtractor/DbRetryProxy.php]: Anyway DEFAULT_RETRIABLE_EXCEPTIONS ;) But to need to change it.

File: tests/DbRetryProxyTest.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    private $logger;
+
+    public function setUp(): void
+    {
+        $this->testHandler = new TestHandler();
+        $this->logger = new Logger('dbRetryProxyTest');
+        $this->logger->pushHandler($this->testHandler);
+    }
+
+    public function testMaxRetries(): void
```
Comment (tomasfejfar) [inline tests/DbRetryProxyTest.php]: #protip if you named the test "what I expect to happen" you'd noticed that there should be more tests. 

`testWillRetryAndSucceedIfMaxRetriesNotReached`
`testWillThrowAfterDefaultNumberOfRetriesReached`
`testCanOverrideDefaultNumberOfRetries`
`testCanOverrideDefaultTimeout`
`testMaxRetriesIsIgnoredWhenCustomRetryPolicyIsUsed`
... ;)

**Pull request merged**